lighthouse/tests/beacon_node.rs

```diff
@@ -1264,47 +1263,47 @@ fn network_load_flag() {
 // Tests for ENR flags.
 #[test]
 fn enr_udp_port_flag() {
-    let port = unused_udp4_port().expect("Unable to find unused port.");
+    let port = 0;
```

Comment:
This is a bit of a philosophical issue, but enr related ports should in principle never be zero, since they are supposed to be externally reachable addresses. It would better align with what this means to use any other number but zero, as this suggests the "let the OS pick one for me" scenario, which this definitely is not

Reply:
The non-zero ports selected by `unused_port` are one of the main sources of CI failures, because we've realised `unused_port` is very prone to racing (if there are _any_ other processes or tests not using `unused_port` then the abstraction breaks completely). Maybe we need to find a way to tweak these tests so that they still test something useful after binding on 0. Not binding on 0 at this point isn't really an option

Reply:
Ohhh, this port isn't actually bound is it. @jmcph4 maybe we can just use a dummy value here, as the port is only used in the ENR

Reply:
None of these tests actually pass -- I'm unsure as to why CI isn't failing at this point. Working on these now.

Reply:
no these ports are not bound. This should not affect it at all, in hindsight I have no idea what that function was used here at all

Reply:
I think it's pretty clear at this stage that you definitely know the code better than I do! I'm just somewhat confused at the moment because (in these tests) each port seems to be treated unilaterally the same (i.e., bound or unbound, ENR or non-ENR, etc.).

Is my following understanding correct?

 - Any port that is to be advertised as part of our ENR needs to be set to a **non-zero** dummy value (let's call these DUMMY_ENR_UDP_PORT`, `DUMMY_ENR_TCP_PORT`, etc.)
 - Any listening port (i.e., a port we'll be binding on) **must** be set to zero (this is to avoid racing when CI inevitably runs these tests)
 - Any port that is to be used by the discovery service to "listen" on must also be set to zero, for the reason immediately above

Reply:
yes, the ports are treated the same right now (before your changes), and that makes no sense. There was no reason why the `unused` functions should have been used for the tests that work on flags that set `enr` values since those are external ports not listening ports.

Your description of the three parts is on point.

In the future, not this pr, the values you have added as zero for enr ports should be rejected, so leaving them here is what you would need to go back and correct. I gave @michaelsproul the following example:

It's like having some code that deals with people's data and adding a test in which someone's age is 5432. You would look at that test funny, then later you would add validation that the data is in a logical range, and then go back to the tests and remove the 5432 for something that doesn't raise an eyebrow (which is what happens when I look at those zeros)

Now, there is the enr-match flag, which is a bit tricky. This is supposed to assume your listening address matches your reachable address. Setting to zero makes sense to get a binding port in the test but if the end result is setting the enr to zero then the behaviour is wrong. This is _another_ issue different from the previous one and should be fixed later too. For those I think you need to either set the zero and deal with the different behaviour afterwards (since it needs to change) or set the ports to a default value. To be explicit, the tricky part about the enr-match tests is that here listening ports and enr ports are supposed to be the same.

Let me know if I can help clarify this some more.

Now, I talked this with @michael too. All the above is the long and boring explanation of why I ask you change the tests that set enr ports (except the match ones) to simply use a dummy value that is not zero. Michael suggested 1111, a nice cute non-zero port number